app/javascript/controllers/polygon_search_controller.js

```diff
@@ -0,0 +1,215 @@
+import { Controller } from "@hotwired/stimulus"
+
+export default class extends Controller {
+  connect() {
+    this.map = this.getMapElement()
+    this.resetBtn = this.getResetButton()
+    // Debounce the onGeojsonChange function so it's triggered at most once per second
+    this.debouncedGeojsonChange = this.debounce(this.onGeojsonChange, 1000)
+
+    this.setupEventListeners()
+  }
+
+  getMapElement() {
+    return document.querySelector("my-map:first-of-type")
+  }
+
+  getResetButton() {
+    return this.map.shadowRoot.querySelector('button[title="Reset map view"]')
+  }
+
+  getCSRFToken() {
+    return document.querySelector("[name='csrf-token']").getAttribute("content")
+  }
+
+  getUrlPath() {
+    return document.querySelector("#os_polygon_path").value
+  }
+
+  getAddressContainer() {
+    return document.getElementById("address-container")
+  }
+
+  getConsultationNeighbourAddressesForm() {
+    return document.getElementById("consultation-neighbour-addresses-form")
+  }
+
+  setupEventListeners() {
+    this.map.addEventListener("geojsonChange", this.debouncedGeojsonChange)
+    this.resetBtn.addEventListener("click", this.clearAddresses)
+  }
+
+  async handleGeojsonChange(geoJSON) {
+    this.clearAddresses()
+    this.showLoadingSpinner()
+
+    try {
+      const response = await this.fetchGeojsonData(geoJSON, this.getCSRFToken())
+      const data = await response.json()
+
+      this.appendAddressesToPage(data)
+    } catch (error) {
+      console.error("There was an error calling the OS Places API:", error)
+    } finally {
+      this.hideLoadingSpinner()
+    }
+  }
+
+  async fetchGeojsonData(geoJSON, csrfToken) {
+    const response = await fetch(this.getUrlPath(), {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        "X-CSRF-Token": csrfToken,
+      },
+      body: JSON.stringify({ geojson: geoJSON }),
+    })
+
+    if (!response.ok) {
+      throw new Error(`HTTP error with status: ${response.status}`)
+    }
+
+    return response
+  }
+
+  onGeojsonChange = ({ detail: geoJSON }) => {
+    const geoJsonFeatures = geoJSON.features
+    if (geoJsonFeatures === undefined) {
+      return
+    }
+
+    this.handleGeojsonChange(geoJsonFeatures[0])
+  }
+
+  clearAddresses = () => {
```

Comment:
Really minor point overall, but would we be better off using the same function syntax where possible? I think these are equivalent in this context.

```suggestion
  clearAddresses() {
```

Reply:
Ha didn't even notice. Yeah better to keep consistent.
Is this something we could add to our js linter 
 Rome ?

Reply:
Just looked and I couldn't find a linter rule for it. (We currently use their recommended set, which expands occasionally when new rules are stable enough, but there doesn't look like any of the experimental rules check for this yet. It may be something they add in the future.)

Reply:
Hm the specs fail when I change to this.

I think it's to do with the interaction with the this keyword.
i think for now I'll leave as is since it works as expected but can make a note to refactor

Reply:
Oh yeah, that makes sense — arrow functions don't get a new `this`. We'd definitely be able to rewrite it but it wouldn't be as simple as I had expected.